### PR TITLE
SOLR-14961:  ZkMaintenanceUtils.clean doesn't remove zk-nodes with same path length

### PR DESCRIPTION
**Pull request by defonion**
# Description

- use ArrayList instead of TreeSet to allow multiple entries that are equal (in this case of equal length)
- sort the ArrayList afterwards

# Solution

Usage of a data stucture that allows to store duplicates in terms of the comparator used.

# Tests

Added test method to SolrZkClientTest, that adds multiple nodes to zookeeper (some of equal length). Afterwards these nodes are deleted via the ZkMaintenanceUtils.clean(...) method. Then the nodes are queried again to check if all nodes that should have been deleted are gone.

# Checklist

Please review the following and check all that apply:

- [ x ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ x ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ x ] I have developed this patch against the `master` branch.
- [ ] I have run `./gradlew check`.
- [ x ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,7 @@ __pycache__
 
 # Emacs backup
 *~
+/.caches/
+/eclipse-build/
+/lucene/
```
Comment (madrob) [inline .gitignore:30]: I... don't think this is right to include `/lucene/` and `/solr/`

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java [hunk 2]
```diff
@@ -259,12 +260,16 @@ public static void clean(SolrZkClient zkClient, String path, Predicate<String> f
       clean(zkClient, path);
       return;
     }
-
-    TreeSet<String> paths = new TreeSet<>(Comparator.comparingInt(String::length).reversed());
+    
+    ArrayList<String> paths = new ArrayList<>();
+//    TreeSet<String> paths = new TreeSet<>(Comparator.comparingInt(String::length).reversed());
```
Comment (madrob) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java:265]: Please don't leave this commented out line in the code.

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java [hunk 3]
```diff
@@ -259,12 +260,16 @@ public static void clean(SolrZkClient zkClient, String path, Predicate<String> f
...
 
     traverseZkTree(zkClient, path, VISIT_ORDER.VISIT_POST, znode -> {
       if (!znode.equals("/") && filter.test(znode)) paths.add(znode);
     });
+    
+    // sort the list in descending order to ensure that child entries are deleted first
```
Comment (madrob) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java:271]: Thank you for including this comment! It's really helpful to me!

File: solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java [hunk 4]
```diff
@@ -220,4 +223,45 @@ public void testCheckInterrupted() {
...
+  public void testClean() throws KeeperException, InterruptedException, SolrServerException, UnsupportedEncodingException {
+    /* PREPARE */
+    String path = "/myPath/isTheBest";
+    String data1 = "myStringData1";
+    String data2 = "myStringData2";
+    String longData = "myLongStringData";
+    // create zk nodes that have the same path length
+    defaultClient.create("/myPath", null, CreateMode.PERSISTENT, true);
+    defaultClient.create(path, null, CreateMode.PERSISTENT, true);
+    defaultClient.create(path +"/file1.txt", data1.getBytes("UTF-8"), CreateMode.PERSISTENT, true);
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java:246]: nit: use StandardCharsets.UTF8

File: solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java [hunk 5]
```diff
@@ -220,4 +223,45 @@ public void testCheckInterrupted() {
...
+    
+    /* RUN */
+    // delete all nodes that contain "file"
+    defaultClient.clean(path, node -> node.contains("file"));
+    
+    /* CHECK */
+    listZnode = defaultClient.listZnode(path, false);
+    log.info(listZnode);
+    // list of node must not contain file1, file2 or some_longer_file2 because they where deleted
+    assertTrue(!listZnode.contains("file1"));
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java:262]: nit: assertFalse

File: solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java [hunk 6]
```diff
@@ -220,4 +223,45 @@ public void testCheckInterrupted() {
...
+   * This test reproduces the issue of trying to delete zk-nodes that have the same length. (SOLR-14961). 
+   * 
+   * @throws InterruptedException when having troublke creating test nodes
+   * @throws KeeperException error when talking to zookeeper
+   * @throws SolrServerException when having trouble connecting to solr 
+   * @throws UnsupportedEncodingException when getBytes() uses unknown encoding
+   * 
+   */
+  @Test
+  public void testClean() throws KeeperException, InterruptedException, SolrServerException, UnsupportedEncodingException {
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java:237]: A better place might be TestZkMaintenanceUtils

**Review by madrob — COMMENTED** [review]
This is a great bug report and PR, thank you for contributing! A few minor notes, but overall it looks really good.

Please add a line to solr/CHANGES.txt with a short summary and your name so that we can credit you - I will make sure it ends up in the right spot when committing.

**Comment by defonion** [comment]
I added the changes recommended by @madrob

**Comment by madrob** [comment]
Hmmm, I wasn't looking closely enough when I made some of those recommendations. You're right that the test should be in `org.apache.solr.common.cloud` but there is also already `o.a.s.common.util.TestZkMaintenanceUtil` - can you combine the two?

File: .gitignore [hunk 7]
```diff
@@ -25,3 +25,6 @@ __pycache__
 
 # Emacs backup
 *~
+/.caches/
+/eclipse-build/
```
Comment (madrob) [inline .gitignore:29]: @dsmiley I don't use Eclipse so I don't know if we need this or not, can you PTAL?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java [hunk 8]
```diff
@@ -259,12 +260,15 @@ public static void clean(SolrZkClient zkClient, String path, Predicate<String> f
...
-    TreeSet<String> paths = new TreeSet<>(Comparator.comparingInt(String::length).reversed());
+    
+    ArrayList<String> paths = new ArrayList<>();
 
     traverseZkTree(zkClient, path, VISIT_ORDER.VISIT_POST, znode -> {
       if (!znode.equals("/") && filter.test(znode)) paths.add(znode);
     });
+    
+    // sort the list in descending order to ensure that child entries are deleted first
+    Collections.sort(paths, Comparator.comparingInt(String::length).reversed());
```
Comment (madrob) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkMaintenanceUtils.java:271]: nit: this should be list.sort

File: solr/solrj/src/test/org/apache/solr/common/cloud/TestZkMaintenanceUtils.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  public void testClean() throws KeeperException, InterruptedException, SolrServerException, UnsupportedEncodingException {
+    /* PREPARE */
+    String path = "/myPath/isTheBest";
+    String data1 = "myStringData1";
+    String data2 = "myStringData2";
+    String longData = "myLongStringData";
+    // create zk nodes that have the same path length
+    defaultClient.create("/myPath", null, CreateMode.PERSISTENT, true);
+    defaultClient.create(path, null, CreateMode.PERSISTENT, true);
+    defaultClient.create(path +"/file1.txt", data1.getBytes("UTF-8"), CreateMode.PERSISTENT, true);
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/common/cloud/TestZkMaintenanceUtils.java:93]: StandardCharsets here too

File: solr/solrj/src/test/org/apache/solr/common/cloud/TestZkMaintenanceUtils.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    zkServer.shutdown();
+    solrClient.close();
+    cluster.shutdown();
+    super.tearDown();
+  }
+
+  /**
+   * This test reproduces the issue of trying to delete zk-nodes that have the same length. (SOLR-14961). 
+   * 
+   * @throws InterruptedException when having troublke creating test nodes
```
Comment (madrob) [inline solr/solrj/src/test/org/apache/solr/common/cloud/TestZkMaintenanceUtils.java:77]: s/troublke/trouble

File: .gitignore (hunk 7 above)
Comment (dsmiley) [reply .gitignore:29]: PTAL?
I don't use Eclipse either.  Let's not modify .gitignore in this PR.

These paths are for previous branches (branch_8x), but not master.  I don't think we should pollute the .gitignore of master branch with directories that should be ignored of previous release branches.  The contributor probably switched branches.  Different checkouts, ideally aided by `git worktree` should be used.

Comment (dsmiley) [reply .gitignore:29]: Maybe a comment at the bottom of this file could raise awareness of this to anyone who is tempted to touch it?

**Pull request closed without merging**